src/Transient/Internals.hs

```diff
@@ -601,264 +600,260 @@ delData x=  modify $ \st -> st{mfData= M.delete (typeOf x ) (mfData st)}
+-- If the maximum number of threads (set with `threads`) has been reached
+-- `parallel` performs the work sequentially in the current thread.
+-- So `parallel` means that 'it can be parallelized if there are any threads
+-- available.'
 --
--- The behaviour of `parallel` depend on `StreamData`; If `SMore`, `parallel` will excute again the
--- IO action. with `SLast`, `SDone` and `SError`, `parallel` will not repeat the IO action anymore.
-parallel  ::    IO (StreamData b) -> TransIO (StreamData b)
-parallel  ioaction= Transient $   do
-    cont <- get                    -- !> "PARALLEL"
-    case event cont of
-         j@(Just _) -> do
-            put cont{event=Nothing}
-            return $ unsafeCoerce j
-         Nothing -> do
-            liftIO $ loop cont ioaction
-            was <- getData `onNothing` return NoRemote
-            when (was /= WasRemote) $ setData WasParallel
-            return Nothing
-
-
--- executes the IO action and then the continuation included in the first parameter
+-- If there is a limit on the number of threads, the thread counter is increased
+-- when the thread finishes, so that future calls to `parallel` can make use of it.
+-- The behaviour of `parallel` depends on `StreamData`:
+-- - If `SMore`, `parallel` will execute the IO action again.
+-- - If `SLast`, `SDone`, or `SError`, `parallel` will finish execution.
+parallel :: IO (StreamData b) -> TransIO (StreamData b)
+parallel action = Transient $ do
+  cont <- get                    -- !> "PARALLEL"
+  case event cont of
+    j@(Just _) -> do
+      put cont { event = Nothing }
+      return $ unsafeCoerce j
+    Nothing -> do
+      liftIO $ loop cont action
+      was <- getData `onNothing` return NoRemote
+      when (was /= WasRemote) $ setData WasParallel
+      return Nothing
+
+-- | Executes the IO action and then the continuation in the first parameter.
 loop :: EventF -> IO (StreamData t) -> IO ()
-loop (cont'@(EventF eff e x fs a b c d _ childs g))  rec  =  do
+loop cont' recursive = do
   chs <- liftIO $ newTVarIO []
-  let cont = EventF eff e x fs a b c d (Just cont') chs g
-      iocont dat= do
-          runStateT (runCont cont) cont{event= Just $ unsafeCoerce dat}
-          return ()
-
-      -- execute the IO computation and then the closure-continuation
-      loop'= forkMaybe False cont $ do
-         mdat <- threadDelay 0 >> rec `catch` \(e :: SomeException) -> return $ SError e
-         case mdat of
-             se@(SError _) ->  iocont se
-             SDone ->          iocont SDone
-             last@(SLast _) -> iocont last
-
-             more@(SMore _) -> do
-                  forkMaybe False cont $ iocont more
-                  loop'
+  let cont = cont' { parent = Just cont', children = chs }
+      iocont dat = do
+        runStateT (runCont cont) cont { event = Just $ unsafeCoerce dat }
+        return ()
+
+      -- Execute the IO computation and then the closure continuation
```

Comment:
Remove second blank line, two blank lines should separate code groups and never used inside function definition.  It's better not to use blank lines inside function definitions at all.  In case of long functions it's better to factor out pieces of functionality into separate functions.